### ISSUE TITLE
[$1000] Task completion message lag- Delayed Real-Time Update for Completed and Reopened Task Messages

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#22111]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login with User A and create a new workspace. Go to settings > workspaces > new workspace
2. Create a room from User A with that workspace in Public mode
3. Go to detail > share code and share the link to User B
4. From User B, open the link and access the room
5. From User A, create a task for User B. After creating the task, enter the task thread
6. From User B, enter the task thread and mark it as complete
7. From User A, reopen the task & from User B complete the task again.
8. Verify if you can see the completed task and reopened task  message in realtime
## Expected Result:
Upon completing the task and reopening it, the task thread should show the updated status and reflect the completed and reopened task messages in real-time. Both User A and User B should be able to see the changes instantly without the need for page refreshes or delays.
## Actual Result:
There is a noticeable lag in the real-time update for the completed and reopened task messages in the task thread. The updates are not reflected immediately, and the changes only appear after performing a page refresh.
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:**
**Reproducible in staging?:**
**Reproducible in production?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/ae328b5e-2114-44f6-a4fe-3752647b22ff

https://github.com/Expensify/App/assets/43996225/f6c23b65-8425-47cd-976e-3662345c4112

**Expensify/Expensify Issue URL:**
**Issue reported by:** @avi-shek-jha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1688174761586429

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01334e50bd6f49ef71</li>
        <li>Upwork Job ID: 1675973780112375808</li>
        <li>Last Price Increase: 2023-07-10</li>
    </ul>
</details>

**Comment by MitchExpensify** [comment]
<img width="1398" alt="image" src="https://github.com/Expensify/App/assets/36425901/ff4791ba-194a-464a-8239-23bcb9fe779d">

**Comment by MitchExpensify** [comment]
I agree that User B  should see a message showing the task was reopened. Its weird that we show User A this message but not User B

**Comment by ygshbht** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The problem we're addressing involves a delay in the real-time update of messages within the task thread when a task is marked as completed or reopened. This leads to situations where the message such as "completed task {task_name}" is not immediately reflected in the chat.

### What is the root cause of that problem?
The core of the problem lies within a combination of the backend and frontend. On one hand, when a task's status is updated, the backend does send pusher/socket events to update the report object data and the count of total messages displayed in the group chat. However, it doesn't send a reportActions_{exampleNum} message which would include the "task completed" or "task reopened" messages.

On the other hand, the frontend is not currently designed to fetch the updated data when a task's status changes. As a result, these status change messages are not displayed in real time in the chat interface.

### What changes do you think we should make in order to solve the problem?
I propose implementing a mechanism on the frontend that checks for changes in the report.state, or in either report.stateNum or report.statusNum. Upon detecting a change, the application should fetch the updated report along with its associated reportActions.

This could be accomplished within the componentDidUpdate lifecycle method as follows:
```
componentDidUpdate(prevProps) {
    if (prevProps.report.state !== this.props.report.state || prevProps.report.stateNum !== this.props.report.stateNum || prevProps.report.statusNum !== this.props.report.statusNum) {
        // Fetch the report with reportActions
        Report.openReport(this.props.report.reportID);
    }
}
```

This modification should ensure that messages related to task status changes are updated in real-time in the chat.

An alternative approach could involve the backend sending the necessary reportActions_{exampleNum} message when a task's status is altered.

**Comment by MitchExpensify** [comment]
How does the proposal above look to you @fedirjh ?

**Comment by MitchExpensify** [comment]
Done, thanks @fedirjh !

**Comment by MitchExpensify** [comment]
Hmm I thought internal would assign an engineer. Applying that via the eng label instead

**Comment by Gonals** [comment]
I'm going to be fairly OOO this next weeks, so rerolling assignment!

**Comment by MitchExpensify** [comment]
Looks like the re-labelling didn't get a new eng assigned, rolling the dice again!

**Comment by MitchExpensify** [comment]
Just to catch you up @arosiclair, @fedirjh's take [is that we'll need some backend work here](https://github.com/Expensify/App/issues/22111#issuecomment-1636277144)

**Comment by MitchExpensify** [comment]
Not overdue, Melvin

**Comment by arosiclair** [comment]
Seems this bug has already been fixed on staging **v1.3.48-0**: 

https://github.com/Expensify/App/assets/6833644/f7eb2877-3b42-400d-9c67-82a89b4efb02

**Comment by arosiclair** [comment]
This is actually a dupe of https://github.com/Expensify/App/issues/21350. Since that report came first there shouldn't be any reporting bonus for this. Closing this out
